### PR TITLE
fix: Prevent panic in Remove Unused Parameter assist

### LINKED ISSUE #12784: "Remove Unused Parameter" panics on nested call
**rust-analyzer version**: 0.0.0 (5342f47f4 2022-07-09)

**rustc version**: rustc 1.62.0 (a8314ef7d 2022-06-27)

When invoking "Removed Unused Parameter", if there is call to the function within the argument list of another call to the function, the request handler panics, failing the `check_disjoint_and_sort(indels)` assert.

```rust
fn foo(x: i32, <|>y: i32) -> i32 {
    x
}

fn bar() {
    foo(1, foo(2, 3));
}
```

Expected:
```rust
fn foo(x: i32) -> i32 {
    x
}

fn bar() {
    foo(1);
}
```

Actual:
request handler panicked: assertion failed: check_disjoint_and_sort(indels)

### PR DESCRIPTION
**Pull request by DorianListens**
Instead of calling `builder.delete` for every text range we find with
`process_usage`, we now ensure that the ranges do not overlap before removing
them. If a range is fully contained by a prior one, it is dropped.

fixes #12784

### CODE REVIEW

**Comment by lnicola** [comment]
As a style nit, I'm not really a fan of `fold` used like this. A `for` loop might look somewhat cleaner.

More importantly, it's somewhat inefficient for a largish number of ranges. I think there's a more efficient way to implement this by sorting the interval endpoints and keeping a list of the open ones, but I don't know if it's worth the complexity.

**Comment by lnicola** [comment]
Something like https://www.baeldung.com/cs/finding-all-overlapping-intervals#sweep-line-approach. But it's probably not something that should block this.

File: crates/ide-assists/src/handlers/remove_unused_param.rs [hunk 1]
```diff
@@ -96,12 +96,20 @@ fn process_usages(
...
     builder.edit_file(file_id);
-    for usage in references {
-        if let Some(text_range) = process_usage(&source_file, usage, arg_to_remove, is_self_present)
-        {
-            builder.delete(text_range);
+    let possible_ranges = references
+        .into_iter()
+        .filter_map(|usage| process_usage(&source_file, usage, arg_to_remove, is_self_present));
+
+    let mut ranges_to_delete: Vec<TextRange> = vec![];
```
Comment (lnicola) [inline crates/ide-assists/src/handlers/remove_unused_param.rs:103]: Nit:

```suggestion
    let mut ranges_to_delete: Vec<TextRange> = Vec::with_capacity(possible_ranges.len());
```

Comment (DorianListens) [reply crates/ide-assists/src/handlers/remove_unused_param.rs:103]: To do this, I'd need to `collect` the `possible_ranges` iterator. I don't know enough about the performance characteristics of `Vec` to know which approach would be more efficient.

Do you think the benefit of `with_capacity` outweighs the potential cost of `collect`?

Comment (lnicola) [reply crates/ide-assists/src/handlers/remove_unused_param.rs:103]: Ah, sorry, I missed that it's an iterator. It's fine to leave it like this.

**Comment by DorianListens** [comment]
I looked into implementing the sweep line style of finding the overlaps, and I think it's definitely doable.
I also think this is something we can easily optimize in the future, so I'm going merge this as is.

@bors r+

**Pull request merged**